client/commands.c

```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
+    processLastBlocksCommand(args, out);
+    return 0;
+}
+
+int xdag_com_keyGen(char * args, FILE* out) {
+    processKeyGenCommand(out);
+    return 0;
+}
+
+int xdag_com_level(char * args, FILE* out) {
+    processLevelCommand(args, out);
+    return 0;
+}
+
+int xdag_com_mining(char * args, FILE* out) {
+    processMiningCommand(args, out);
+    return 0;
+}
+
+int xdag_com_net(char * args, FILE* out) {
+    processNetCommand(args, out);
+    return 0;
+}
+
+int xdag_com_pool(char * args, FILE* out) {
+    processPoolCommand(args, out);
+    return 0;
+}
+
+int xdag_com_miners(char * args, FILE* out) {
+    processMinersCommand(args, out);
+    return 0;
+}
+
+int xdag_com_stats(char * args, FILE* out) {
+    processStatsCommand(out);
+    return 0;
+}
+
+int xdag_com_state(char * args, FILE* out) {
+    fprintf(out, "%s\n", get_state());
+    return 0;
+}
+
+int xdag_com_run(char * args, FILE* out) {
+    g_xdag_run = 1;
+    return 0;
+}
+
+int xdag_com_terminate(char * args, FILE* out) {
+    processExitCommand();
+    return -1;
+}
+
+int xdag_com_exit(char * args, FILE* out) {
+    processExitCommand();
+    return -1;
+}
+
+int xdag_com_help(char *args, FILE* out) {
+    printHelp(out);
+    return 0;
+}
+
+char ** xdag_com_completion(const char *text, int start, int end) {
+    char **matches;
+    matches = (char **)NULL;
+    if (start == 0)
+        matches = rl_completion_matches(text, xdag_com_generator);
+    return (matches);
+}
+
+char* dupstr(char* s){
+    char *r;
+    r = malloc(strlen(s) + 1);
+    strcpy(r, s);
+    return (r);
+}
+
+char* xdag_com_generator(const char* text, int state) {
+    static int list_index, len;
+    char *name;
+    if (!state) {
+        list_index = 0;
+        len = strlen(text);
+    }
+    
+    while ((name = commands[list_index].name)) {
+        list_index++;
+        if (strncmp(name, text, len) == 0)
+            return (dupstr(name));
+    }
+    return ((char *)NULL);
+}
+
+XDAG_COMMAND* find_xdag_command(char *name) {
+    register int i;
+    for (i = 0; commands[i].name; i++)
+        if (strcmp(name, commands[i].name) == 0)
+            return (&commands[i]);
+    return ((XDAG_COMMAND *)NULL);
+}
+
 void startCommandProcessing(int transportFlags)
 {
-	char cmd[XDAG_COMMAND_MAX] = {0};
-
-	if(!(transportFlags & XDAG_DAEMON)) printf("Type command, help for example.\n");
-	for(;;) {
-		if(transportFlags & XDAG_DAEMON) sleep(100);
-		else {
-			printf("%s> ", g_progname);
-			fflush(stdout);
-			fgets(cmd, XDAG_COMMAND_MAX, stdin);
-			if(xdag_command(cmd, stdout) < 0) {
-				break;
-			}
-		}
-	}
+    char *cmd;
+    if(!(transportFlags & XDAG_DAEMON)) printf("Type command, help for example.\n");
+    
+    // readline lib
+    rl_readline_name = "xdag";
+    rl_attempted_completion_function = xdag_com_completion;
+    
+    for (;;) {
+        if(transportFlags & XDAG_DAEMON) {
+            sleep(100);
+        } else {
+            cmd = readline("xdag> ");
```

Comment:
Do you know if this function works with Windows? Could you please check and add precompilation directives if necessary?
like 
#if WIN32
printf("%s> ", g_progname);
fflush(stdout);
fgets(cmd, XDAG_COMMAND_MAX, stdin);
#else
cmd = readline("xdag> ");
#endif

Reply:
good suggest